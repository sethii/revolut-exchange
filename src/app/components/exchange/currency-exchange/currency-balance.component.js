import React from 'react';
import styled from 'react-emotion';
import PropTypes from 'prop-types';
import { setPropTypes } from 'recompose';
import getSymbolFromCurrency from 'currency-symbol-map';
import { twoDecimalPlacesFormatter } from '../../../utils';

const enhance = setPropTypes({
  noBalance: PropTypes.bool.isRequired,
  wallet: PropTypes.shape({
    code: PropTypes.string.isRequired,
    amount: PropTypes.number.isRequired
  }).isRequired
});

const Balance = styled.p`
  font-size: 12px;
  color: ${props => props.noBalance ? 'rgba(247, 0, 135, 1)' : '#b1b1b1'};
  font-family: Arial;
  
  & span {
    font-size: 10px;
  }
`;

export const CurrencyBalance = enhance(({ noBalance, wallet }) => <Balance noBalance={noBalance}>
  Balance: <span>{ getSymbolFromCurrency(wallet.code) }</span>{ twoDecimalPlacesFormatter(wallet.amount)}
</Balance>);
