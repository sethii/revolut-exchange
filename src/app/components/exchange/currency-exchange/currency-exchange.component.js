import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import { setPropTypes, compose, defaultProps } from 'recompose';
import { ExchangeFee } from '../exchange-fee/exchange-fee.component';
import { CurrencySelect } from './currency-select.component';
import { CurrencyBalance } from './currency-balance.component';
import { CurrencyInput } from './currency-input.component';

const enhance = compose(
  setPropTypes({
    amount: PropTypes.string.isRequired,
    currencies: PropTypes.arrayOf(
      PropTypes.shape({
        code: PropTypes.string.isRequired
      })
    ).isRequired,
    selectedCurrency: PropTypes.shape({
        code: PropTypes.string.isRequired
      }).isRequired,
    onCurrencyChange: PropTypes.func.isRequired,
    onAmountChange: PropTypes.func.isRequired,
    wallet: PropTypes.shape({
      code: PropTypes.string.isRequired
    }),
    prefix: PropTypes.string.isRequired,
    type: PropTypes.oneOf(['from', 'to']).isRequired,
    fee: PropTypes.number,
    openFeeInfo: PropTypes.func,
    noBalance: PropTypes.bool
  }),
  defaultProps({
    noBalance: false,
    fee: 0
  })
);

const Container = styled.div`
  width: 600px;
  display: flex;
  margin: 0 auto;
  align-self: ${(props) => props.type === 'from' ? 'flex-end' : 'flex-start'};
  padding: 20px 0px;
`;

const Column = styled.div`
  width: 50%;
  display: flex;
  flex-direction: column;
  align-items: ${(props) => props.position === 'end' ? 'flex-end' : 'flex-start'};
`;

export const CurrencyExchange = enhance((props) => {
  return <Container type={props.type}>
    <Column position={'start'}>
      <CurrencySelect currencies={props.currencies} selectedCurrency={props.selectedCurrency} onCurrencyChange={props.onCurrencyChange} />
      <CurrencyBalance noBalance={props.noBalance} wallet={props.wallet}/>
    </Column>
    <Column position={'end'}>
      <CurrencyInput amount={props.amount} prefix={props.prefix} onChange={props.onAmountChange}/>
      { props.type === 'to' && <ExchangeFee
        fee={props.fee}
        currency={props.selectedCurrency}
        openFeeInfo={props.openFeeInfo}
      />}
    </Column>
  </Container>
});
