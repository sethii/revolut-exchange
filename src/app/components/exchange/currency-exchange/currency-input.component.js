import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import { compose, setPropTypes, withHandlers } from 'recompose';
import { numberFormatter, prefixedNumberFormatter } from '../../../utils';

const handleSupportedKeys = (event, currentValue) => {
  if (event.keyCode === 8) {
    if (currentValue.toString().length === 0) {
      return currentValue
    }

    return numberFormatter(currentValue.toString().substr(0, currentValue.toString().length - 1))
  }

  return numberFormatter(currentValue + event.key);
};

const enhance = compose(
  withHandlers({
    onKeyDown: (props) => event => {
      const isValidCharacter = /[\d.,]/.test(event.key);
      const isSecondDot = (event.key === '.' || event.key === ',') && /[.,]/.test(props.amount);
      const isOnlyDot = (event.key === '.' || event.key === ',') && props.amount.length === 0;

      if ((!isValidCharacter || isSecondDot || isOnlyDot) && event.keyCode !== 8) {
        event.preventDefault();
      } else {
        const currentValue = handleSupportedKeys(event, props.amount);
        props.onChange(currentValue);
      }
    }
  }),
  setPropTypes({
    amount: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    prefix: PropTypes.string.isRequired,
    onKeyDown: PropTypes.func.isRequired
  })
);

const TextInput = styled.input`
  outline: none;
  border: none;
  padding: 10px 0px;
  font-size: 20px;
  text-align: right;
  user-select: none;
  background-color: transparent;
  
  & ::selection {
    background: transparent;
  }
`;

export const CurrencyInput = enhance(({ amount, prefix, onKeyDown }) => <TextInput
  type="text"
  value={prefixedNumberFormatter(amount, prefix)}
  placeholder="0"
  onKeyDown={onKeyDown}
  onChange={() => {}}
/>);
