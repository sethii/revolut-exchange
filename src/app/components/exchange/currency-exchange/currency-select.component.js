import React from 'react';
import styled from 'react-emotion';
import PropTypes from 'prop-types';
import { setPropTypes } from 'recompose';

const Select = styled.select`
  height: 43px;
  width: 80px;
  outline: none;
`;

const enhance = setPropTypes({
  currencies: PropTypes.array.isRequired,
  selectedCurrency: PropTypes.shape({
    code: PropTypes.string.isRequired,
  }).isRequired,
  onCurrencyChange: PropTypes.func.isRequired
});

export const CurrencySelect = enhance(({ currencies, selectedCurrency, onCurrencyChange}) => <Select
  value={currencies.findIndex((currency) => currency.code === selectedCurrency.code)}
  onChange={(event) => onCurrencyChange(currencies[event.target.value])}
>
    { currencies.map((currency, index) => <option key={index} value={index}>{currency.code}</option>) }
</Select>);
