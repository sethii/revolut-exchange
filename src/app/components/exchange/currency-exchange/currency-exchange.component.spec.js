import React from 'react';
import renderer from 'react-test-renderer';
import { toClass } from 'recompose';

import { CurrencyExchange } from './currency-exchange.component';

describe('Currency exchange component', () => {
  it('renders exchange component for from currency', () => {
    const amount= '1234';
    const wallet = {
      code: 'USD',
      amount: 1000
    };
    const onAmountChange = jest.fn();
    const prefix = "+";
    const onCurrencyChange = jest.fn();
    const currencies = [
      {
        code: 'USD',
      },
      {
        code: 'PLN'
      }
    ];
    const selectedCurrency = {
      code: 'PLN'
    };
    const type = 'from';
    const fee = 50;
    const openFeeInfo = jest.fn();
    const noBalance = false;

    const CurrencyExchangeComponent = toClass(CurrencyExchange);

    expect(renderer.create(<CurrencyExchangeComponent
      amount={amount}
      prefix={prefix}
      wallet={wallet}
      onAmountChange={onAmountChange}
      onCurrencyChange={onCurrencyChange}
      currencies={currencies}
      selectedCurrency={selectedCurrency}
      type={type}
      fee={fee}
      openFeeInfo={openFeeInfo}
      noBalance={noBalance}
    />).toJSON()).toMatchSnapshot();
  });

  it('renders exchange component for to currency', () => {
    const amount= '1234';
    const wallet = {
      code: 'USD',
      amount: 1000
    };
    const onAmountChange = jest.fn();
    const prefix = "+";
    const onCurrencyChange = jest.fn();
    const currencies = [
      {
        code: 'USD',
      },
      {
        code: 'PLN'
      }
    ];
    const selectedCurrency = {
      code: 'PLN'
    };
    const type = 'to';
    const fee = 50;
    const openFeeInfo = jest.fn();
    const noBalance = false;

    const CurrencyExchangeComponent = toClass(CurrencyExchange);

    expect(renderer.create(<CurrencyExchangeComponent
      amount={amount}
      prefix={prefix}
      wallet={wallet}
      onAmountChange={onAmountChange}
      onCurrencyChange={onCurrencyChange}
      currencies={currencies}
      selectedCurrency={selectedCurrency}
      type={type}
      fee={fee}
      openFeeInfo={openFeeInfo}
      noBalance={noBalance}
    />).toJSON()).toMatchSnapshot();
  });
});
