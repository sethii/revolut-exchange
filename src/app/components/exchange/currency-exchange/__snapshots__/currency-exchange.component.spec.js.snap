// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency exchange component renders exchange component for from currency 1`] = `
.glamor-11 {
  width: 600px;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  margin: 0 auto;
  -ms-flex-item-align: flex-end;
  align-self: flex-end;
  padding: 20px 0px;
}

.glamor-5 {
  width: 50%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.glamor-1 {
  height: 43px;
  width: 80px;
  outline: none;
}

.glamor-3 {
  font-size: 12px;
  color: #b1b1b1;
  font-family: Arial;
}

.glamor-3 span {
  font-size: 10px;
}

.glamor-9 {
  width: 50%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.glamor-7 {
  outline: none;
  border: none;
  padding: 10px 0px;
  font-size: 20px;
  text-align: right;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
}

.glamor-7 ::-moz-selection {
  background: transparent;
}

.glamor-7 ::selection {
  background: transparent;
}

<div
  className="glamor-10 glamor-11"
  type="from"
>
  <div
    className="glamor-4 glamor-5"
  >
    <select
      className="glamor-0 glamor-1"
      onChange={[Function]}
      value={1}
    >
      <option
        value={0}
      >
        USD
      </option>
      <option
        value={1}
      >
        PLN
      </option>
    </select>
    <p
      className="glamor-2 glamor-3"
    >
      Balance: 
      <span>
        $
      </span>
      1000
    </p>
  </div>
  <div
    className="glamor-4 glamor-9"
  >
    <input
      className="glamor-6 glamor-7"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="0"
      type="text"
      value="+1234"
    />
  </div>
</div>
`;

exports[`Currency exchange component renders exchange component for to currency 1`] = `
.glamor-5 {
  width: 50%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.glamor-1 {
  height: 43px;
  width: 80px;
  outline: none;
}

.glamor-3 {
  font-size: 12px;
  color: #b1b1b1;
  font-family: Arial;
}

.glamor-3 span {
  font-size: 10px;
}

.glamor-11 {
  width: 50%;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.glamor-7 {
  outline: none;
  border: none;
  padding: 10px 0px;
  font-size: 20px;
  text-align: right;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
}

.glamor-7 ::-moz-selection {
  background: transparent;
}

.glamor-7 ::selection {
  background: transparent;
}

.glamor-13 {
  width: 600px;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  margin: 0 auto;
  -ms-flex-item-align: flex-start;
  align-self: flex-start;
  padding: 20px 0px;
}

.glamor-9 {
  font-size: 12px;
  color: #b1b1b1;
  font-family: Arial;
}

.glamor-9 span {
  font-size: 10px;
}

.glamor-9 i:hover {
  cursor: pointer;
}

<div
  className="glamor-12 glamor-13"
  type="to"
>
  <div
    className="glamor-4 glamor-5"
  >
    <select
      className="glamor-0 glamor-1"
      onChange={[Function]}
      value={1}
    >
      <option
        value={0}
      >
        USD
      </option>
      <option
        value={1}
      >
        PLN
      </option>
    </select>
    <p
      className="glamor-2 glamor-3"
    >
      Balance: 
      <span>
        $
      </span>
      1000
    </p>
  </div>
  <div
    className="glamor-4 glamor-11"
  >
    <input
      className="glamor-6 glamor-7"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="0"
      type="text"
      value="+1234"
    />
    <p
      className="glamor-8 glamor-9"
    >
      Inc. fee: 
      <span>
        z≈Ç
      </span>
      50
       
      <i
        aria-hidden="true"
        className="fa fa-info-circle"
        onClick={[Function]}
      />
    </p>
  </div>
</div>
`;
